File: src/free_proxy/infrastructure/network/routing.py
from __future__ import annotations

import asyncio
import logging
import sys
from typing import Protocol

from free_proxy.config import Settings
from free_proxy.domain.exceptions import NetworkOperationError
from free_proxy.infrastructure.network.commands import CommandResult, SystemCommandRunner

logger = logging.getLogger(__name__)


class CommandRunner(Protocol):
    async def run(self, command: list[str], *, timeout_seconds: float = 5) -> CommandResult: ...


class PolicyRouter:
    def __init__(
        self,
        settings: Settings,
        runner: CommandRunner | None = None,
        platform: str | None = None,
    ) -> None:
        self._settings = settings
        self._runner = runner or SystemCommandRunner()
        self._platform = platform or sys.platform
        self._rp_filter_original: dict[str, str] = {}

    @property
    def supported(self) -> bool:
        return self._platform.startswith("linux")

    async def setup(self, interface: str | None = None) -> None:
        if not self.supported:
            raise NetworkOperationError("Policy routing is only supported on Linux")
        device = interface or self._settings.tunnel_interface
        table = str(self._settings.policy_routing_table)
        last_error: NetworkOperationError | None = None
        for attempt in range(1, self._settings.routing_setup_retries + 1):
            try:
                await self._setup_once(device, table)
                return
            except NetworkOperationError as exc:
                last_error = exc
                await self.cleanup()
                if attempt < self._settings.routing_setup_retries:
                    await asyncio.sleep(self._settings.routing_retry_interval_seconds)
        assert last_error is not None
        raise last_error

    async def _setup_once(self, device: str, table: str) -> None:
        await self.cleanup()

        route_result = await self._runner.run(
            ["ip", "route", "add", "default", "dev", device, "table", table]
        )
        if route_result.returncode != 0:
            raise NetworkOperationError(
                "Unable to add policy route for "
                f"{device}: command={' '.join(['ip', 'route', 'add'])} "
                f"returncode={route_result.returncode} stderr={route_result.stderr.strip()}"
            )

        rule_result = await self._runner.run(["ip", "rule", "add", "oif", device, "table", table])
        if rule_result.returncode != 0:
            await self.cleanup()
            raise NetworkOperationError(
                f"Unable to add policy rule for {device}: {rule_result.stderr.strip()}"
            )

        for target in ("all", "default", device):
            read_result = await self._runner.run(
                ["sysctl", "-n", f"net.ipv4.conf.{target}.rp_filter"]
            )
            if read_result.returncode == 0:
                self._rp_filter_original[target] = read_result.stdout.strip()
            result = await self._runner.run(
                ["sysctl", "-w", f"net.ipv4.conf.{target}.rp_filter=2"]
            )
            if result.returncode != 0:
                message = f"Unable to configure rp_filter for {target}: {result.stderr.strip()}"
                if self._settings.routing_strict_rp_filter:
                    raise NetworkOperationError(message)
                logger.warning(message)

    async def cleanup(self) -> None:
        if not self.supported:
            return
        table = str(self._settings.policy_routing_table)
        await self._runner.run(["ip", "rule", "del", "table", table])
        await self._runner.run(["ip", "route", "flush", "table", table])
        for target, value in self._rp_filter_original.items():
            await self._runner.run(["sysctl", "-w", f"net.ipv4.conf.{target}.rp_filter={value}"])
        self._rp_filter_original.clear()

File: internal/netx/routing.go
package netx

import (
	"context"
	"fmt"
	"log/slog"
	"runtime"
	"strconv"
	"strings"
	"time"
)

// PolicyRouter installs the policy route/rule that forces marked traffic through
// the tunnel and relaxes rp_filter, restoring prior values on cleanup. Linux only.
type PolicyRouter struct {
	runner        CommandRunner
	table         int
	iface         string
	setupRetries  int
	retryInterval time.Duration
	strictRPF     bool

	rpFilterOriginal map[string]string
}

// PolicyRouterConfig carries the settings a PolicyRouter needs.
type PolicyRouterConfig struct {
	Table         int
	Interface     string
	SetupRetries  int
	RetryInterval time.Duration
	StrictRPF     bool
}

// NewPolicyRouter constructs a PolicyRouter.
func NewPolicyRouter(runner CommandRunner, cfg PolicyRouterConfig) *PolicyRouter {
	if runner == nil {
		runner = SystemCommandRunner{}
	}
	if cfg.SetupRetries < 1 {
		cfg.SetupRetries = 1
	}
	return &PolicyRouter{
		runner:           runner,
		table:            cfg.Table,
		iface:            cfg.Interface,
		setupRetries:     cfg.SetupRetries,
		retryInterval:    cfg.RetryInterval,
		strictRPF:        cfg.StrictRPF,
		rpFilterOriginal: map[string]string{},
	}
}

// Supported reports whether policy routing is available on this OS.
func (r *PolicyRouter) Supported() bool { return runtime.GOOS == "linux" }

// Setup installs the route/rule with retries, cleaning up between attempts.
func (r *PolicyRouter) Setup(ctx context.Context, iface string) error {
	if !r.Supported() {
		return fmt.Errorf("policy routing is only supported on Linux")
	}
	device := iface
	if device == "" {
		device = r.iface
	}
	table := strconv.Itoa(r.table)
	var lastErr error
	for attempt := 1; attempt <= r.setupRetries; attempt++ {
		if err := r.setupOnce(ctx, device, table); err != nil {
			lastErr = err
			_ = r.Cleanup(ctx)
			if attempt < r.setupRetries && r.retryInterval > 0 {
				time.Sleep(r.retryInterval)
			}
			continue
		}
		return nil
	}
	return lastErr
}

func (r *PolicyRouter) setupOnce(ctx context.Context, device, table string) error {
	_ = r.Cleanup(ctx)

	routeRes, err := r.runner.Run(ctx, []string{"ip", "route", "add", "default", "dev", device, "table", table}, 5*time.Second)
	if err != nil {
		return err
	}
	if routeRes.ReturnCode != 0 {
		return fmt.Errorf("unable to add policy route for %s: %s", device, strings.TrimSpace(routeRes.Stderr))
	}
	ruleRes, err := r.runner.Run(ctx, []string{"ip", "rule", "add", "oif", device, "table", table}, 5*time.Second)
	if err != nil {
		return err
	}
	if ruleRes.ReturnCode != 0 {
		_ = r.Cleanup(ctx)
		return fmt.Errorf("unable to add policy rule for %s: %s", device, strings.TrimSpace(ruleRes.Stderr))
	}
	for _, target := range []string{"all", "default", device} {
		key := "net.ipv4.conf." + target + ".rp_filter"
		if read, err := r.runner.Run(ctx, []string{"sysctl", "-n", key}, 5*time.Second); err == nil && read.ReturnCode == 0 {
			r.rpFilterOriginal[target] = strings.TrimSpace(read.Stdout)
		}
		set, err := r.runner.Run(ctx, []string{"sysctl", "-w", key + "=2"}, 5*time.Second)
		if err != nil || set.ReturnCode != 0 {
			msg := fmt.Sprintf("unable to configure rp_filter for %s", target)
			if r.strictRPF {
				return fmt.Errorf("%s", msg)
			}
			slog.Warn(msg, "module", "netx")
		}
	}
	return nil
}

// Cleanup removes the route/rule and restores rp_filter values.
func (r *PolicyRouter) Cleanup(ctx context.Context) error {
	if !r.Supported() {
		return nil
	}
	table := strconv.Itoa(r.table)
	_, _ = r.runner.Run(ctx, []string{"ip", "rule", "del", "table", table}, 5*time.Second)
	_, _ = r.runner.Run(ctx, []string{"ip", "route", "flush", "table", table}, 5*time.Second)
	for target, value := range r.rpFilterOriginal {
		_, _ = r.runner.Run(ctx, []string{"sysctl", "-w", "net.ipv4.conf." + target + ".rp_filter=" + value}, 5*time.Second)
	}
	r.rpFilterOriginal = map[string]string{}
	return nil
}
